lib/IRGen/GenMeta.cpp

```diff
@@ -3837,13 +3839,21 @@ namespace {
   public:
     ForeignClassMetadataBuilder(IRGenModule &IGM, ClassDecl *target,
                                 ConstantStructBuilder &B)
-      : ForeignMetadataBuilderBase(IGM, target, B) {}
+        : ForeignMetadataBuilderBase(IGM, target, B) {
+      if (IGM.getOptions().LazyInitializeClassMetadata)
+        CanBeConstant = false;
+    }
 
     void emitInitializeMetadata(IRGenFunction &IGF, llvm::Value *metadata,
                                 MetadataDependencyCollector *collector) {
       // Emit a reference to the superclass.
-      auto superclass = IGF.emitAbstractTypeMetadataRef(
-                                   Target->getSuperclass()->getCanonicalType());
+      llvm::Value *superclass;
+      if (auto parent = Target->getSuperclass())
+        superclass = IGF.emitAbstractTypeMetadataRef(parent->getCanonicalType());
+      else if (auto parent = (IGM.ObjCInterop
+                                 ? IGM.Context.TheUnknownObjectType
+                                 : IGM.Context.TheNativeObjectType))
+        superclass = IGF.emitAbstractTypeMetadataRef(parent->getCanonicalType());
```

Comment:
I don't think its right to set the superclass to the metadata for one of these types?

Reply:
Isn't this the metaclass setup?

Reply:
I'm not sure what you mean by that. If you look at what the code did before your change, the superclass field was null if there was no superclass. You're setting it to Builtin.NativeObject/UnknownObject, which are not class types, therefore cannot appear in the superclass field of a class.

Reply:
Right, this logic still seems bizarre and busted.

Reply:
Yeah, I got nothing.